File: Source/Core/Common/Time.vb
' MIT License
'
' Copyright (c) 2017 Robert Lodico
'
' Permission is hereby granted, free of charge, to any person obtaining a copy
' of this software and associated documentation files (the "Software"), to deal
' in the Software without restriction, including without limitation the rights
' to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
' copies of the Software, and to permit persons to whom the Software is
' furnished to do so, subject to the following conditions:
'
' The above copyright notice and this permission notice shall be included in all
' copies or substantial portions of the Software.
'
' THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
' IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
' FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
' AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
' LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
' OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
' SOFTWARE.
Imports System.Timers

Public Enum TimeOfDay As Byte
    None
    Day
    Night
    Dawn
    Dusk
End Enum

Public Delegate Sub HandleTimeEvent(ByRef source As Time)

Public Class Time
    Private Shared _mInstance As Time = Nothing

    Public Shared ReadOnly Property Instance As Time
        Get
            If (_mInstance Is Nothing) Then
                _mInstance = New Time()
            End If

            Return _mInstance
        End Get
    End Property

    Public Event OnTimeChanged As HandleTimeEvent
    Public Event OnTimeOfDayChanged As HandleTimeEvent
    Public Event OnTimeSync As HandleTimeEvent

    Private ReadOnly _mTimer As Timer

    Private _mTime As Date

    Public Property Time As Date
        Get
            Return _mTime
        End Get
        Set
            _mTime = Value

            Dim newTimeOfDay As TimeOfDay = GetTimeOfDay(Time.Hour)
            If (TimeOfDay <> newTimeOfDay) Then TimeOfDay = newTimeOfDay

            RaiseEvent OnTimeChanged(Me)
        End Set
    End Property

    Private _mGameSpeed As Double

    Public Property GameSpeed As Double
        Get
            Return _mGameSpeed
        End Get
        Set
            _mGameSpeed = Value
            RaiseEvent OnTimeSync(Me)
        End Set
    End Property

    Private mSyncInterval As Integer

    Public Property SyncInterval As Integer
        Get
            Return mSyncInterval
        End Get
        Set
            mSyncInterval = Value

            _mTimer.Stop()
            _mTimer.Interval = mSyncInterval
            _mTimer.Start()
            RaiseEvent OnTimeSync(Me)
        End Set
    End Property

    Private mTimeOfDay As TimeOfDay

    Public Property TimeOfDay As TimeOfDay
        Get
            Return mTimeOfDay
        End Get
        Set
            mTimeOfDay = Value
            RaiseEvent OnTimeOfDayChanged(Me)
        End Set
    End Property

    Public Sub New()
        mSyncInterval = 6000.0

        _mTimer = New Timer(SyncInterval)

        AddHandler _mTimer.Elapsed, AddressOf HandleTimerElapsed

        _mTimer.Start()
    End Sub

    Private Sub HandleTimerElapsed(sender As Object, e As ElapsedEventArgs)
        RaiseEvent OnTimeSync(Me)
    End Sub

    Public Overrides Function ToString() As String
        Return ToString("h:mm:ss tt")
    End Function

    Public Overloads Function ToString(ByRef format As String) As String
        Return Time.ToString(format)
    End Function

    Public Sub Reset()
        Time = New DateTime(0)
    End Sub

    Public Sub Tick()
        Time = Time.AddSeconds(GameSpeed)
    End Sub

    Public Shared Function GetTimeOfDay(ByRef hours As Integer) As TimeOfDay
        If (hours < 6) Then
            Return TimeOfDay.Night
        ElseIf (6 <= hours AndAlso hours <= 9) Then
            Return TimeOfDay.Dawn
        ElseIf (9 < hours AndAlso hours < 18) Then
            Return TimeOfDay.Day
        Else
            Return TimeOfDay.Dusk
        End If
    End Function
End Class

File: Source/Core/Time.cs
using System;
// MIT License
// 
// Copyright (c) 2017 Robert Lodico
// 
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
// 
// The above copyright notice and this permission notice shall be included in all
// copies or substantial portions of the Software.
// 
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
using System.Timers;

namespace Core
{

    public enum TimeOfDay : byte
    {
        None = 0,
        Day,
        Night,
        Dawn,
        Dusk
    }

    public delegate void HandleTimeEvent(ref Clock source);

    public class Clock
    {
        private static Clock _mInstance = null;

        public static Clock Instance
        {
            get
            {
                if (_mInstance is null)
                {
                    _mInstance = new Clock();
                }

                return _mInstance;
            }
        }

        public event HandleTimeEvent OnTimeChanged;
        public event HandleTimeEvent OnTimeOfDayChanged;
        public event HandleTimeEvent OnTimeSync;

        private readonly Timer _mTimer;

        private DateTime _mTime;

        public DateTime Time
        {
            get
            {
                return _mTime;
            }
            set
            {
                _mTime = value;

                int arghours = Time.Hour;
                var newTimeOfDay = GetTimeOfDay(ref arghours);
                if (TimeOfDay != newTimeOfDay)
                    TimeOfDay = newTimeOfDay;

                var argsource = this;
                OnTimeChanged?.Invoke(ref argsource);
            }
        }

        private double _mGameSpeed;

        public double GameSpeed
        {
            get
            {
                return _mGameSpeed;
            }
            set
            {
                _mGameSpeed = value;
                var argsource = this;
                OnTimeSync?.Invoke(ref argsource);
            }
        }

        private int mSyncInterval;

        public int SyncInterval
        {
            get
            {
                return mSyncInterval;
            }
            set
            {
                mSyncInterval = value;

                _mTimer.Stop();
                _mTimer.Interval = mSyncInterval;
                _mTimer.Start();
                var argsource = this;
                OnTimeSync?.Invoke(ref argsource);
            }
        }

        private TimeOfDay mTimeOfDay;

        public TimeOfDay TimeOfDay
        {
            get
            {
                return mTimeOfDay;
            }
            set
            {
                mTimeOfDay = value;
                var argsource = this;
                OnTimeOfDayChanged?.Invoke(ref argsource);
            }
        }

        public Clock()
        {
            mSyncInterval = (int)Math.Round(6000.0);

            _mTimer = new Timer(SyncInterval);

            _mTimer.Elapsed += HandleTimerElapsed;

            _mTimer.Start();
        }

        private void HandleTimerElapsed(object sender, ElapsedEventArgs e)
        {
            var argsource = this;
            OnTimeSync?.Invoke(ref argsource);
        }

        public override string ToString()
        {
            string argformat = "h:mm:ss tt";
            return ToString(ref argformat);
        }

        public string ToString(ref string format)
        {
            return Time.ToString(format);
        }

        public void Reset()
        {
            Time = new DateTime(0L);
        }

        public void Tick()
        {
            Time = Time.AddSeconds(GameSpeed);
        }

        public static TimeOfDay GetTimeOfDay(ref int hours)
        {
            if (hours < 6)
            {
                return TimeOfDay.Night;
            }
            else if (6 <= hours & hours <= 9)
            {
                return TimeOfDay.Dawn;
            }
            else if (9 < hours & hours < 18)
            {
                return TimeOfDay.Day;
            }
            else
            {
                return TimeOfDay.Dusk;
            }
        }
    }
}